File: examples/go/node_template_example.go
package main

import (
	"github.com/castai/pulumi-castai/sdk/go/castai" // Assuming main package
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// runNodeTemplateExample demonstrates creating a node template.
// Renamed from main to avoid conflicts.
func runNodeTemplateExample(ctx *pulumi.Context) error {
	// Placeholder: Assume these IDs are obtained from other resources
	clusterId := pulumi.String("your-cluster-id")
	nodeConfigurationId := pulumi.String("your-node-config-id")

	nodeTemplate, err := castai.NewNodeTemplate(ctx, "example-go-node-template", &castai.NodeTemplateArgs{
		ClusterId:       clusterId,
		Name:            pulumi.String("default-go-template"),
		IsEnabled:       pulumi.Bool(true),
		ConfigurationId: nodeConfigurationId,
		ShouldTaint:     pulumi.Bool(true),
		CustomLabels: pulumi.StringMap{
			"environment":    pulumi.String("production"),
			"pulumi-managed": pulumi.String("true"),
		},
		CustomTaints: castai.NodeTemplateCustomTaintArray{
			&castai.NodeTemplateCustomTaintArgs{
				Key:    pulumi.String("pulumi-dedicated"),
				Value:  pulumi.String("backend"),
				Effect: pulumi.String("NoSchedule"),
			},
		},
		Constraints: &castai.NodeTemplateConstraintsArgs{
			OnDemand:                               pulumi.Bool(true),
			Spot:                                   pulumi.Bool(false),
			UseSpotFallbacks:                       pulumi.Bool(true),
			FallbackRestoreRateSeconds:             pulumi.Int(300),
			EnableSpotDiversity:                    pulumi.Bool(true),
			SpotDiversityPriceIncreaseLimitPercent: pulumi.Int(20),
			SpotInterruptionPredictionsEnabled:     pulumi.Bool(true),
			SpotInterruptionPredictionsType:        pulumi.String("aws-rebalance-recommendations"), // Or other types
			ComputeOptimizedState:                  pulumi.String("disabled"),                      // enabled, disabled
			StorageOptimizedState:                  pulumi.String("disabled"),                      // enabled, disabled
			IsGpuOnly:                              pulumi.Bool(false),
			MinCpu:                                 pulumi.Int(2),
			MaxCpu:                                 pulumi.Int(8),
			MinMemory:                              pulumi.Int(4096),                                                             // MiB
			MaxMemory:                              pulumi.Int(16384),                                                            // MiB
			Architectures:                          pulumi.StringArray{pulumi.String("amd64")},                                   // , "arm64"
			Azs:                                    pulumi.StringArray{pulumi.String("us-east-1a"), pulumi.String("us-east-1b")}, // Adjust AZs
			BurstableInstances:                     pulumi.String("disabled"),                                                    // enabled, disabled
			// CustomerSpecific:                   pulumi.String("disabled"), // Needs specific config if enabled
			InstanceFamilies: &castai.NodeTemplateConstraintsInstanceFamiliesArgs{
				Include: pulumi.StringArray{pulumi.String("m5"), pulumi.String("c5")}, // Example families
				// Exclude: pulumi.StringArray{},
			},
			// CustomPriority: &castai.NodeTemplateConstraintsCustomPriorityArgs{...}, // Optional
		},
		// IsDefault:       pulumi.Bool(false), // Only one template can be default
	})
	if err != nil {
		return err
	}

	// Export the template ID
	ctx.Export("node_template_go_id", nodeTemplate.ID())

	return nil
}

/*
// Example of how you might call this from your main function:
func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        if err := runNodeTemplateExample(ctx); err != nil {
            return err
        }
        // ... potentially call other example functions ...
        return nil
    })
}
*/

File: examples/typescript/node_template_example.ts
import * as pulumi from "@pulumi/pulumi";
import * as castai from "@pulumi/castai"; // Use your actual package name

// Placeholder: Assume these IDs are obtained from other resources
const clusterId = "your-cluster-id";
const nodeConfigurationId = "your-node-config-id";

const nodeTemplate = new castai.NodeTemplate("example-ts-node-template", {
    clusterId: clusterId,
    name: "default-ts-template",
    isEnabled: true,
    configurationId: nodeConfigurationId,
    shouldTaint: true,
    customLabels: {
        "environment": "production",
        "pulumi-managed": "true",
    },
    customTaints: [{
        key: "pulumi-dedicated",
        value: "backend",
        effect: "NoSchedule",
    }],
    constraints: {
        onDemand: true,
        spot: false,
        useSpotFallbacks: true,
        fallbackRestoreRateSeconds: 300,
        enableSpotDiversity: true,
        spotDiversityPriceIncreaseLimitPercent: 20,
        spotInterruptionPredictionsEnabled: true,
        spotInterruptionPredictionsType: "aws-rebalance-recommendations", // Or other types
        computeOptimizedState: "disabled", // enabled, disabled
        storageOptimizedState: "disabled", // enabled, disabled
        isGpuOnly: false,
        minCpu: 2,
        maxCpu: 8,
        minMemory: 4096, // MiB
        maxMemory: 16384, // MiB
        architectures: ["amd64"], // , "arm64"
        azs: ["us-east-1a", "us-east-1b"], // Adjust AZs
        burstableInstances: "disabled", // enabled, disabled
        // customerSpecific: "disabled", // Needs specific config if enabled
        instanceFamilies: {
            include: ["m5", "c5"], // Example families
            // exclude: []
        },
        // customPriority: { ... } // Optional
    },
    // isDefault: false, // Only one template can be default
});

// Export the template ID
export const nodeTemplateId = nodeTemplate.id; 